source/opt/unify_const.cpp

```diff
@@ -0,0 +1,169 @@
+ResultIdTrieNode* GetOrCreateIteratorFor(
+    uint64_t key,
+    std::unordered_map<uint64_t, std::unique_ptr<ResultIdTrieNode>>* m) {
+  auto iter = m->find(key);
+  if (iter == m->end()) {
+    // insert new entry
+    (*m)[key] = std::unique_ptr<ResultIdTrieNode>(new ResultIdTrieNode);
+    iter = m->find(key);
+  }
+  return iter->second.get();
+}
+
+// A helper function to construct the keys used to identify a constant. The
+// keys are consisent of 1) opcode, 2) result type id, 3) all other operands.
+// Constants with same keys are guaranteed to be exactly the same.
+std::vector<uint64_t> GetLookUpKeys(const ir::Instruction& inst) {
+  std::vector<uint64_t> keys;
+  // Need to use the opcode, otherwise there might be a conflict with the
+  // following case when <op>'s binary value equals xx's id:
+  //  OpSpecConstantOp tt <op> yy zz
+  //  OpSpecConstantComposite tt xx yy zz;
+  keys.push_back(static_cast<uint64_t>(inst.opcode()));
+  for (uint32_t i = 0; i < inst.NumOperands(); i++) {
+    if (inst.GetOperand(i).type == SPV_OPERAND_TYPE_RESULT_ID) continue;
+    keys.push_back(GetOperandValueInUint64(inst.GetOperand(i)));
+  }
+  return keys;
+}
+
+// Tries to find the same constant defining instruction of the given |inst| in
+// the trie pointed by the given trie node |node|. If a same instruction is
+// found in the trie, returns its result id. Otherwise, creates an trie node
+// for the given instruction then returns its result id.
+uint32_t LookupOrInsert(ResultIdTrieNode* node, const ir::Instruction& inst) {
+  // get an array of keys to lookup in the trie nodes.
+  std::vector<uint64_t> keys = GetLookUpKeys(inst);
+  // use the keys to lookup/update the trie nodes.
+  for (uint64_t key : keys) {
+    node = GetOrCreateIteratorFor(key, &node->next);
+  }
+  if (node->result_id == 0) {
+    node->result_id = inst.result_id();
+  }
+  return node->result_id;
+}
+
+// Removes the OpName instructions and OpDecorate instructions that are
+// referencing the given instruction.
+void RemoveOpNameAndOpDecorateUsersOfInstruction(
+    const ir::Instruction& inst, analysis::DefUseManager* def_use_mgr) {
+  std::unordered_set<ir::Instruction*> opname_and_opdecorate_to_remove;
+  for (auto u : *def_use_mgr->GetUses(inst.result_id())) {
+    switch (u.inst->opcode()) {
+      case SpvOp::SpvOpName:
+      case SpvOp::SpvOpDecorate:
+        opname_and_opdecorate_to_remove.insert(u.inst);
+      default:
+        break;
+    }
+  }
+  for (auto* i : opname_and_opdecorate_to_remove) {
+    def_use_mgr->KillInst(i);
+  }
+}
+}  // anonymous namespace
+
+bool UnifyConstantPass::Process(ir::Module* module) {
+  bool modified = false;
+  ResultIdTrieNode defined_constants;
+  std::unordered_set<ir::Instruction*> duplicated_cosnts;
+  analysis::DefUseManager def_use_mgr(module);
+
+  for (ir::Instruction& inst : module->types_values()) {
+    switch (inst.opcode()) {
+      case SpvOp::SpvOpConstantTrue:
```

Comment:
Add also mention that `SpecId` should not applied to `SpecConstant{Composite|Op}`, so it's ok to hash them here.